# Reglas de asociación y market-basket analysis

Las reglas de asociación permiten encontrar patrones comunes en grandes conjuntos de datos.

En este caso para identificar los productos que se compra de forma conjunta en una tienda.

Proceso:

Instalamos la libreria MLxtend con "pip install mlxtend" y importamos en nuestro notebook para trabajar con los algoritmos de Apriori y Reglas de Asociacion junto con Pandas

In [2]:
import pandas as pd
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules

Importamos nuestra base de datos

In [35]:
df = pd.read_csv("factura_final_mediana.csv")
df.head(30)

Codigo                                      Producto  Unidades  Precio  \
0     1119                                 Fruta al peso     1.110    1.05   
1     1035                             Cerveza 33cl SKOL     6.000    3.00   
2     1031              Cerveza sin 0,0 33cl SAN MIGUEL      1.000    0.70   
3     1060                              Agua 1,5L BEZOYA     1.000    0.70   
4      421              Tomate frito 265gr brick ORLANDO     1.000    0.50   
5     1067                         Cerveza 1l SAN MIGUEL     1.000    1.50   
6     1078                                  Chuches 0.05    20.000    1.00   
7     1072                   Agua garrafa 5l FONT NATURA     2.000    1.80   
8     1119                                 Fruta al peso     0.245    0.47   
9     1066                            Agua 1,5L LANJARON     1.000    0.70   
10     903   Cafe helado para beber  vaso 230ml LANDESSA     1.000    1.30   
11     904     Cafe helado para beber vaso 250ml  KALISE     1.000    1.30   
12     904     Cafe helado para beber vaso 250ml  KALISE     1.000    1.30   
13    1116                            Kinder bueno white     1.000    1.00   
14     950  Queso tierno bajo en sal 80gr GARCIA BAQUERO     1.000    1.00   
15     950  Queso tierno bajo en sal 80gr GARCIA BAQUERO     1.000    1.00   
16    1072                   Agua garrafa 5l FONT NATURA     1.000    0.90   
17    1072                   Agua garrafa 5l FONT NATURA     1.000    0.90   
18    1028                          Aquarius limon 1,5L      1.000    2.00   
19    1028                          Aquarius limon 1,5L      1.000    2.00   
20    1116                            Kinder bueno white     1.000    1.00   
21    1062                            Agua 1,5L FONTAREL     1.000    0.70   
22     927                        Bacalao palito ALBACOR     1.000    1.99   
23     927                        Bacalao palito ALBACOR     1.000    1.99   
24    1062                            Agua 1,5L FONTAREL     1.000    0.70   
25    1062                            Agua 1,5L FONTAREL     1.000    0.70   
26    1119                                 Fruta al peso     0.240    0.94   
27    1119                                 Fruta al peso     0.640    1.22   
28    1119                                 Fruta al peso     0.230    0.38   
29    1119                                 Fruta al peso     0.410    0.64   

    Num_Factura  Precio_Unit  Total   Coste  Seccion           Fecha_Hora  \
0             1         0.95   1.05  0.7600      7.0  2015-08-16 22:43:00   
1             2         0.50   3.00  0.2800     10.0  2015-08-16 22:59:00   
2             3         0.70   1.90  0.3990     10.0  2015-08-20 22:39:00   
3             3         0.70   1.90  0.3570     10.0  2015-08-20 22:39:00   
4             3         0.50   1.90  0.3000      7.0  2015-08-20 22:39:00   
5             4         1.50   2.50  1.0050     10.0  2015-08-20 22:56:00   
6             4         0.05   2.50  0.0400      7.0  2015-08-20 22:56:00   
7             5         0.90   1.80  0.5310     10.0  2015-08-21 08:48:00   
8             6         1.90   0.47  1.5200      7.0  2015-08-21 09:20:59   
9             7         0.70   2.00  0.3780     10.0  2015-08-21 09:38:59   
10            7         1.30   2.00  0.7670     10.0  2015-08-21 09:38:59   
11            8         1.30   2.80  0.7540     10.0  2015-08-21 09:42:00   
12            8         1.30   4.90  0.7540     10.0  2015-08-21 08:53:00   
13            8         1.00   4.90  0.6800      7.0  2015-08-21 08:53:00   
14            8         1.00   2.80  0.7500      7.0  2015-08-21 09:42:00   
15            8         1.00   4.90  0.7500      7.0  2015-08-21 08:53:00   
16            8         0.90   2.80  0.5310     10.0  2015-08-21 09:42:00   
17            8         0.90   4.90  0.5310     10.0  2015-08-21 08:53:00   
18            8         2.00   2.80  1.4000      1.0  2015-08-21 09:42:00   
19            8         2.00   4.90  1.4000  

Para hacer el modelos,hay que eliminar los espacios que continene en las descripciones contienen espacios que ha de ser eliminados. 

In [4]:
df['Producto'] = df['Producto'].str.strip()

Creamos una variable "basket" con una query que esta filtrada previamente por las bebidas, con el Numero de factura y productos por sus unidades. Cada una de las filas representa una factura y cada una de las columnas un producto, e indica el numero resultante indica el total de unidades de cada producto vendido en cada factura.

In [5]:
basket = (df.query("Seccion==88").groupby(['Num_Factura', 'Producto'])['Unidades'].sum().unstack().reset_index().fillna(0).set_index('Num_Factura'))

In [6]:
basket

Producto     AGUA 8L MIJAS  AGUA ERMITICA 1.5L  AGUA SOLAN DE CABRAS 1.5L  \
Num_Factura                                                                 
2                      0.0                 0.0                        0.0   
3                      0.0                 0.0                        0.0   
4                      0.0                 0.0                        0.0   
5                      0.0                 0.0                        0.0   
7                      0.0                 0.0                        0.0   
...                    ...                 ...                        ...   
467102                 0.0                 0.0                        0.0   
467104                 0.0                 0.0                        0.0   
467106                 0.0                 0.0                        0.0   
467107                 0.0                 0.0                        0.0   
467110                 0.0                 0.0                        0.0   

Producto     AGUA0.5LEL CAÑAR  AMSTEL RADLER  Agua 0.5L fuente primavera  \
Num_Factura                                                                
2                         0.0            0.0                         0.0   
3                         0.0            0.0                         0.0   
4                         0.0            0.0                         0.0   
5                         0.0            0.0                         0.0   
7                         0.0            0.0                         0.0   
...                       ...            ...                         ...   
467102                    0.0            0.0                         0.0   
467104                    0.0            0.0                         0.0   
467106                    0.0            0.0                         0.0   
467107                    0.0            0.0                         0.0   
467110                    0.0            0.0                         0.0   

Producto     Agua 1,5L BEZOYA  Agua 1,5L FONT VELLA  Agua 1,5L FONTAREL  \
Num_Factura                                                               
2                         0.0                   0.0                 0.0   
3                         1.0                   0.0                 0.0   
4                         0.0                   0.0                 0.0   
5                         0.0                   0.0                 0.0   
7                         0.0                   0.0                 0.0   
...                       ...                   ...                 ...   
467102                    0.0                   0.0                 0.0   
467104                    0.0                   0.0                 0.0   
467106                    0.0                   0.0                 0.0   
467107                    0.0                   0.0                 0.0   
467110                    0.0                   0.0                 0.0   

Producto     Agua 1,5L LANJARON  ...  \
Num_Factura                      ...   
2                           0.0  ...   
3                           0.0  ...   
4                           0.0  ...   
5                           0.0  ...   
7                           1.0  ...   
...                         ...  ...   
467102                      0.0  ...   
467104                      0.0  ...   
467106                      0.0  ...   
467107                      0.0  ...   
467110                      0.0  ...   

Producto     tinto de verano la casera Limón Sin Alcohol  \
Num_Factura                                                
2                                                    0.0   
3                                                    0.0   
4                                                    0.0   
5                                                    0.0   
7                                                    0.0   
...                                                  ...   
467102                          

Se define una funcion que da valores 0 si el valor es igual a 0, o 1 en caso contrario y se aplica a la variable basket previamente creada

In [17]:
def encode_units(x):
    if x == 0:
        return 0
    if x > 0:
        return 1

In [8]:
basket_sets = basket.applymap(encode_units)

In [9]:
basket_sets

Producto     AGUA 8L MIJAS  AGUA ERMITICA 1.5L  AGUA SOLAN DE CABRAS 1.5L  \
Num_Factura                                                                 
2                        0                   0                          0   
3                        0                   0                          0   
4                        0                   0                          0   
5                        0                   0                          0   
7                        0                   0                          0   
...                    ...                 ...                        ...   
467102                   0                   0                          0   
467104                   0                   0                          0   
467106                   0                   0                          0   
467107                   0                   0                          0   
467110                   0                   0                          0   

Producto     AGUA0.5LEL CAÑAR  AMSTEL RADLER  Agua 0.5L fuente primavera  \
Num_Factura                                                                
2                           0              0                           0   
3                           0              0                           0   
4                           0              0                           0   
5                           0              0                           0   
7                           0              0                           0   
...                       ...            ...                         ...   
467102                      0              0                           0   
467104                      0              0                           0   
467106                      0              0                           0   
467107                      0              0                           0   
467110                      0              0                           0   

Producto     Agua 1,5L BEZOYA  Agua 1,5L FONT VELLA  Agua 1,5L FONTAREL  \
Num_Factura                                                               
2                           0                     0                   0   
3                           1                     0                   0   
4                           0                     0                   0   
5                           0                     0                   0   
7                           0                     0                   0   
...                       ...                   ...                 ...   
467102                      0                     0                   0   
467104                      0                     0                   0   
467106                      0                     0                   0   
467107                      0                     0                   0   
467110                      0                     0                   0   

Producto     Agua 1,5L LANJARON  ...  \
Num_Factura                      ...   
2                             0  ...   
3                             0  ...   
4                             0  ...   
5                             0  ...   
7                             1  ...   
...                         ...  ...   
467102                        0  ...   
467104                        0  ...   
467106                        0  ...   
467107                        0  ...   
467110                        0  ...   

Producto     tinto de verano la casera Limón Sin Alcohol  \
Num_Factura                                                
2                                                      0   
3                                                      0   
4                                                      0   
5                                                      0   
7                                                      0   
...                                                  ...   
467102                          

La función "apriori" es una función de asociación de ítems que se utiliza para encontrar conjuntos de ítems que ocurren con cierta frecuencia en un conjunto de datos.

El "min_support" es un valor numérico que especifica el umbral mínimo de soporte que deben tener los conjuntos de ítems para ser considerados frecuentes. 

Da lugar a un dataframe con los articulos que se repiten con ese min_support aplicado.

In [10]:
frequent_itemsets = apriori(basket_sets, min_support=0.001, use_colnames=True)
frequent_itemsets

/Users/franciscomanueljimenezpalacios/opt/anaconda3/lib/python3.9/site-packages/mlxtend/frequent_patterns/fpcommon.py:111: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


support                                   itemsets
0    0.009323                       (AGUA ERMITICA 1.5L)
1    0.001512                (AGUA SOLAN DE CABRAS 1.5L)
2    0.024545                         (Agua 1,5L BEZOYA)
3    0.003477                     (Agua 1,5L FONT VELLA)
4    0.003217                       (Agua 1,5L FONTAREL)
..        ...                                        ...
134  0.002339      (Coca cola 33cl, Coca cola zero 33cl)
135  0.001346         (Fanta limon 33cl, Coca cola 33cl)
136  0.002947       (Fanta naranja 33cl, Coca cola 33cl)
137  0.001159  (Fanta naranja 33cl, Coca cola zero 33cl)
138  0.001206     (Fanta naranja 33cl, Fanta limon 33cl)

[139 rows x 2 columns]

La función de asociación de ítems que se utiliza para generar reglas de asociación a partir de conjuntos de ítems frecuentes.
La métrica "lift", que mide la fuerza de la asociación entre dos ítems.
"min_threshold" especifica el umbral mínimo que deben tener las reglas de asociación para ser incluidas en el resultado.

In [11]:
rules = association_rules(frequent_itemsets, metric="lift", min_threshold=1)
rules

antecedents  \
0                      (Agua 1,5L LANJARON)   
1                          (Coca cola 33cl)   
2            (Cerveza 33cl crital HEINEKEN)   
3                (Agua 1,5L pack6 LANJARON)   
4               (Alhambra Tradicional 50cl)   
5                     (cervesa alhambra 1l)   
6                    (Cerveza 1l CRUZCAMPO)   
7                  (Cerveza 33cl CRUZCAMPO)   
8                       (Cerveza 33cl SKOL)   
9                         (Cerveza 1l SKOL)   
10                  (Cerveza 33cl VICTORIA)   
11                (Cerveza 33cl SAN MIGUEL)   
12                  (Cerveza 33cl VICTORIA)   
13                      (Cerveza 33cl SKOL)   
14  (Coca Cola Lata 330ml Zero/sin cafeina)   
15                      (Cerveza 33cl SKOL)   
16                               (skol 500)   
17                      (Cerveza 33cl SKOL)   
18           (Cerveza 33cl crital HEINEKEN)   
19              (Cerveza Alhambra botellin)   
20                           (Coca cola 2L)   
21                       (Fanta naranja 2L)   
22                         (Coca cola 33cl)   
23                    (Coca cola zero 33cl)   
24                       (Fanta limon 33cl)   
25                         (Coca cola 33cl)   
26                     (Fanta naranja 33cl)   
27                         (Coca cola 33cl)   
28                     (Fanta naranja 33cl)   
29                    (Coca cola zero 33cl)   
30                     (Fanta naranja 33cl)   
31                       (Fanta limon 33cl)   

                                consequents  antecedent support  \
0                          (Coca cola 33cl)            0.016854   
1                      (Agua 1,5L LANJARON)            0.067077   
2                (Agua 1,5L pack6 LANJARON)            0.010389   
3            (Cerveza 33cl crital HEINEKEN)            0.009911   
4                     (cervesa alhambra 1l)            0.019270   
5               (Alhambra Tradicional 50cl)            0.030911   
6                  (Cerveza 33cl CRUZCAMPO)            0.039102   
7                    (Cerveza 1l CRUZCAMPO)            0.047391   
8                         (Cerveza 1l SKOL)            0.065060   
9                       (Cerveza 33cl SKOL)            0.038146   
10                (Cerveza 33cl SAN MIGUEL)            0.048025   
11                  (Cerveza 33cl VICTORIA)            0.010934   
12                      (Cerveza 33cl SKOL)            0.048025   
13                  (Cerveza 33cl VICTORIA)            0.065060   
14                      (Cerveza 33cl SKOL)            0.015269   
15  (Coca Cola Lata 330ml Zero/sin cafeina)            0.065060   
16                      (Cerveza 33cl SKOL)            0.008118   
17                               (skol 500)            0.065060   
18              (Cerveza Alhambra botellin)            0.010389   
19           (Cerveza 33cl crital HEINEKEN)            0.008674   
20                       (Fanta naranja 2L)            0.043374   
21                           (Coca cola 2L)            0.013346   
22                    (Coca cola zero 33cl)            0.067077   
23                         (Coca cola 33cl)            0.031743   
24                         (Coca cola 33cl)            0.010342   
25                       (Fanta limon 33cl)            0.067077   
26                         (Coca cola 33cl)            0.022492   
27                     (Fanta naranja 33cl)            0.067077   
28                    (Coca cola zero 33cl)            0.022492   
29                     (Fanta naranja 33cl)            0.031743   
30                       (Fanta limon 33cl)            0.022492   
31                     (Fanta naranja 33cl)            0.010342   

    consequent support   support  confidence       lift  leverage  conviction  
0             0.067077  0.001169    0.069380   1.034339  0.000039    1.002475  
1             0.016854  0.001169    0.017432   1.034339  0.000039    1.000589  
2             0.009911  0.00

Support (Soporte) mide el porcentaje de la fuerza en la que ocurre la relacion entre el antecedente y la consecuencia.
El Agua 1,5L LANJARON y la Coca cola 33cl aparecen juntos en los tickets en un 0.1169% de las veces.

La confidence (Confianza) es el porcentaje de las transacciones en las que aparece el antecedente en la que también aparece el consecuente. Lo que mide este indicador es la fiabilidad de la regla.

Antecedente o consecuente indica cual es el que tiene mas peso en el acompañamiento

La mejora de la confianza es la fracción de soporte de la regla respecto al que se observaría en caso de independía.

# Analisis con todos los productos

Eliminas el producto de Bolsas porque no es representativo en el estudio y al ser de lo mas vendido desvirtua un poco los datos

In [12]:
df.drop(df.query("Producto == 'Bolsa Plastico LEY'").index,axis=0,inplace=True)

Filtramos por el año 2021 con todos los productos

In [24]:
df1=df[(df['Fecha_Hora'] > '2021-01-01') & (df['Fecha_Hora'] < '2021-12-31')]

In [25]:
#df1['Producto'] = df1['Producto'].str.strip()
#df1.dropna(axis=0, subset=['Num_Factura'], inplace=True)
#df1['Num_Factura'] = df1['Num_Factura'].astype('str')


/var/folders/md/0kp4d7755rz9llrkjzrjm5000000gn/T/ipykernel_1527/1566963137.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df1['Producto'] = df1['Producto'].str.strip()
/var/folders/md/0kp4d7755rz9llrkjzrjm5000000gn/T/ipykernel_1527/1566963137.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df1.dropna(axis=0, subset=['Num_Factura'], inplace=True)
/var/folders/md/0kp4d7755rz9llrkjzrjm5000000gn/T/ipykernel_1527/1566963137.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer]

In [26]:
basket1 = (df1.groupby(['Num_Factura', 'Producto'])['Unidades'].sum().unstack().reset_index().fillna(0).set_index('Num_Factura'))

In [27]:
basket_sets1 = basket1.applymap(encode_units)
frequent_itemsets1 = apriori(basket_sets1, min_support=0.002, use_colnames=True)

/Users/franciscomanueljimenezpalacios/opt/anaconda3/lib/python3.9/site-packages/mlxtend/frequent_patterns/fpcommon.py:111: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


In [28]:
rules1 = association_rules(frequent_itemsets1, metric="lift", min_threshold=1)
rules1.head(60)

antecedents                          consequents  \
0                       (Fruta al peso)                         (1/2 huevos)   
1                          (1/2 huevos)                      (Fruta al peso)   
2                                 (Pan)                         (1/2 huevos)   
3                          (1/2 huevos)                                (Pan)   
4                                 (Pan)                          (12 huevos)   
5                           (12 huevos)                                (Pan)   
6                                 (Pan)        (Agua garrafa 8l FONT NATURA)   
7         (Agua garrafa 8l FONT NATURA)                                (Pan)   
8    (Blue Chameleon Bebida Energetica)                  (Cerveza 33cl SKOL)   
9                   (Cerveza 33cl SKOL)   (Blue Chameleon Bebida Energetica)   
10   (Blue Chameleon Bebida Energetica)                       (Chuches 0.05)   
11                       (Chuches 0.05)   (Blue Chameleon Bebida Energetica)   
12                 (Producto de Oferta)                  (Bocadillos oferta)   
13                  (Bocadillos oferta)                 (Producto de Oferta)   
14                            (Caña Qe)                 (Producto de Oferta)   
15                 (Producto de Oferta)                            (Caña Qe)   
16               (Cerveza 1l CRUZCAMPO)             (Cerveza 33cl CRUZCAMPO)   
17             (Cerveza 33cl CRUZCAMPO)               (Cerveza 1l CRUZCAMPO)   
18                                (Pan)     (Chorizo tradicion 85gr REVILLA)   
19     (Chorizo tradicion 85gr REVILLA)                                (Pan)   
20                       (Chuches 0.05)                 (Producto de Oferta)   
21                 (Producto de Oferta)                       (Chuches 0.05)   
22                (Hielo bolsa cubitos)                       (Coca cola 2L)   
23                       (Coca cola 2L)                (Hielo bolsa cubitos)   
24                       (Coca cola 2L)                 (Producto de Oferta)   
25                 (Producto de Oferta)                       (Coca cola 2L)   
26                 (Producto de Oferta)                     (Coca cola 33cl)   
27                     (Coca cola 33cl)                 (Producto de Oferta)   
28                 (Producto de Oferta)                (Coca cola zero 33cl)   
29                (Coca cola zero 33cl)                 (Producto de Oferta)   
30                  (Embutido al corte)                      (Fruta al peso)   
31                      (Fruta al peso)                  (Embutido al corte)   
32                                (Pan)                  (Embutido al corte)   
33                  (Embutido al corte)                                (Pan)   
34                  (Embutido al corte)                 (Producto de Oferta)   
35                 (Producto de Oferta)                  (Embutido al corte)   
36                      (Fruta al peso)         (PACK AGUA 1.5L FONT NATURA)   
37         (PACK AGUA 1.5L FONT NATURA)                      (Fruta al peso)   
38                                (Pan)                      (Fruta al peso)   
39                      (Fruta al peso)                                (Pan)   
40      (Pechuga de pavo 100gr EL POZO)                      (Fruta al peso)   
41                      (Fruta al peso)      (Pechuga de pavo 100gr EL POZO)   
42                      (Fruta al peso)                 (Producto de Oferta)   
43                 (Producto de Oferta)                      (Fruta al peso)   
44                      (Fruta al peso)                    (lechuga iceberg)   
45                    (lechuga iceberg)                      (Fruta al peso)   
46                                (Pan)       (Jamon Curado 8u 55gr NAVIDUL)   
47       (Jamon Curado 8u 55gr NAVIDUL)                                (Pan)   
48                                (Pan)         (Jamon cocido 100gr EL POZO)   
49         (Jamon cocido 100

Filtramos por el año 2016 con todos los productos. Primer año completo

In [31]:
df2=df[(df['Fecha_Hora'] > '2016-01-01') & (df['Fecha_Hora'] < '2016-12-31')]
basket2 = (df2.groupby(['Num_Factura', 'Producto'])['Unidades'].sum().unstack().reset_index().fillna(0).set_index('Num_Factura'))
basket_sets2 = basket2.applymap(encode_units)
frequent_itemsets2 = apriori(basket_sets2, min_support=0.002, use_colnames=True)
rules2 = association_rules(frequent_itemsets2, metric="lift", min_threshold=1)
rules2.head(60)

/Users/franciscomanueljimenezpalacios/opt/anaconda3/lib/python3.9/site-packages/mlxtend/frequent_patterns/fpcommon.py:111: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


antecedents  \
0                                  (Fruta al peso)   
1                                     (1/2 huevos)   
2                                            (Pan)   
3                                     (1/2 huevos)   
4                             (Producto de Oferta)   
5                                     (1/2 huevos)   
6                                  (Fruta al peso)   
7                                      (12 huevos)   
8                                            (Pan)   
9                                      (12 huevos)   
10                            (AGUA ERMITICA 1.5L)   
11                            (Producto de Oferta)   
12                              (Agua 1,5L BEZOYA)   
13                            (Producto de Oferta)   
14                            (Agua 1,5L LANJARON)   
15                            (Producto de Oferta)   
16                                           (Pan)   
17                   (Agua garrafa 8l FONT NATURA)   
18                             (Bocadillos oferta)   
19                                (Coca cola 33cl)   
20                             (Bocadillos oferta)   
21                            (Producto de Oferta)   
22                         (Cerveza 33cl VICTORIA)   
23                             (Cerveza 33cl SKOL)   
24                                           (Pan)   
25                (Chorizo tradicion 85gr REVILLA)   
26                                  (Chuches 0.05)   
27                            (Producto de Oferta)   
28                                           (Pan)   
29                                  (Coca cola 2L)   
30                            (Producto de Oferta)   
31                                (Coca cola 33cl)   
32                                           (Pan)   
33                          (Coca cola light 33cl)   
34                            (Producto de Oferta)   
35                           (Coca cola zero 33cl)   
36                             (Embutido al corte)   
37                                 (Fruta al peso)   
38                                           (Pan)   
39                             (Embutido al corte)   
40                             (Embutido al corte)   
41                            (Producto de Oferta)   
42                                           (Pan)   
43                              (Fanta naranja 2L)   
44                            (Fanta naranja 33cl)   
45                                           (Pan)   
46                            (Fanta naranja 33cl)   
47                            (Producto de Oferta)   
48                  (Jamon Curado 8u 55gr NAVIDUL)   
49                                 (Fruta al peso)   
50                                 (Fruta al peso)   
51                          (Leche Omega 3 PULEVA)   
52                                           (Pan)   
53                                 (Fruta al peso)   
54                 (Pechuga de pavo 100gr EL POZO)   
55                                 (Fruta al peso)   
56                                 (Fruta al peso)   
57                            (Producto de Oferta)   
58                                 (Fruta al peso)   
59  (Queso tierno bajo en sal 80gr GARCIA BAQUERO)   

                                       consequents  antecedent support  \
0                                     (1/2 huevos)            0.216774   
1                                  (Fruta al peso)            0.019730   
2                                     (1/2 huevos)            0.257498   
3                                            (Pan)            0.019730   
4                                     (1/2 huevos)            0.116302   
5                             (Producto de Oferta)            0.019730   
6                                      (12 huevos)            0.216774   
7                                  (Fruta al peso)            0.009404   
8                                      (12 huevos)            0.257498   
9   

Filtramos por el año 2020 con todos los productos. Año de la pandemia

In [33]:
df3=df[(df['Fecha_Hora'] > '2020-01-01') & (df['Fecha_Hora'] < '2020-12-31')]
basket3 = (df3.groupby(['Num_Factura', 'Producto'])['Unidades'].sum().unstack().reset_index().fillna(0).set_index('Num_Factura'))
basket_sets3 = basket3.applymap(encode_units)
frequent_itemsets3 = apriori(basket_sets3, min_support=0.002, use_colnames=True)
rules3 = association_rules(frequent_itemsets3, metric="lift", min_threshold=1)
rules3.head(60)

/Users/franciscomanueljimenezpalacios/opt/anaconda3/lib/python3.9/site-packages/mlxtend/frequent_patterns/fpcommon.py:111: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


antecedents                         consequents  \
0                      (Fruta al peso)                        (1/2 huevos)   
1                         (1/2 huevos)                     (Fruta al peso)   
2                                (Pan)                        (1/2 huevos)   
3                         (1/2 huevos)                               (Pan)   
4                      (Fruta al peso)                         (12 huevos)   
5                          (12 huevos)                     (Fruta al peso)   
6                                (Pan)                         (12 huevos)   
7                          (12 huevos)                               (Pan)   
8                      (Fruta al peso)       (Agua garrafa 5l FONT NATURA)   
9        (Agua garrafa 5l FONT NATURA)                     (Fruta al peso)   
10                               (Pan)       (Agua garrafa 5l FONT NATURA)   
11       (Agua garrafa 5l FONT NATURA)                               (Pan)   
12                     (Fruta al peso)       (Agua garrafa 8l FONT NATURA)   
13       (Agua garrafa 8l FONT NATURA)                     (Fruta al peso)   
14                               (Pan)       (Agua garrafa 8l FONT NATURA)   
15       (Agua garrafa 8l FONT NATURA)                               (Pan)   
16  (Blue Chameleon Bebida Energetica)                      (Chuches 0.05)   
17                      (Chuches 0.05)  (Blue Chameleon Bebida Energetica)   
18                 (Bocadillos oferta)                    (Coca cola 33cl)   
19                    (Coca cola 33cl)                 (Bocadillos oferta)   
20                (Producto de Oferta)                 (Bocadillos oferta)   
21                 (Bocadillos oferta)                (Producto de Oferta)   
22                               (Pan)                           (Caña Qe)   
23                           (Caña Qe)                               (Pan)   
24                           (Caña Qe)                (Producto de Oferta)   
25                (Producto de Oferta)                           (Caña Qe)   
26                 (Cerveza 33cl SKOL)                   (Cerveza 1l SKOL)   
27                   (Cerveza 1l SKOL)                 (Cerveza 33cl SKOL)   
28                               (Pan)    (Chorizo tradicion 85gr REVILLA)   
29    (Chorizo tradicion 85gr REVILLA)                               (Pan)   
30     (Salchichon extra 85gr REVILLA)    (Chorizo tradicion 85gr REVILLA)   
31    (Chorizo tradicion 85gr REVILLA)     (Salchichon extra 85gr REVILLA)   
32                      (Chuches 0.05)               (Gusanitos 0,35 RISI)   
33               (Gusanitos 0,35 RISI)                      (Chuches 0.05)   
34                      (Chuches 0.05)                (Producto de Oferta)   
35                (Producto de Oferta)                      (Chuches 0.05)   
36                      (Coca cola 2L)                (Producto de Oferta)   
37                (Producto de Oferta)                      (Coca cola 2L)   
38                (Producto de Oferta)                    (Coca cola 33cl)   
39                    (Coca cola 33cl)                (Producto de Oferta)   
40                (Producto de Oferta)               (Coca cola zero 33cl)   
41               (Coca cola zero 33cl)                (Producto de Oferta)   
42                               (Pan)            (Cola Cola Zero Cafeina)   
43            (Cola Cola Zero Cafeina)                               (Pan)   
44                               (Pan)       (Donuts Blanco Oferta Pack 4)   
45       (Donuts Blanco Oferta Pack 4)                               (Pan)   
46                 (Embutido al corte)                     (Fruta al peso)   
47                     (Fruta al peso)                 (Embutido al corte)   
48                               (Pan)                 (Embutido al corte)   
49                 (Embutido al corte)                               (Pan)   
50                 (Embutido al corte)           